src/comments.ts

```diff
@@ -0,0 +1,297 @@
+    if (remoteUrl === '') {
+        throw new Error('Git repository has no remote url configured')
+    }
+    const data = await queryGraphQL(
+        gql`
+            query DiscussionThreads(
+                $targetRepositoryGitCloneURL: String!
+                $targetRepositoryPath: String!
+                $relativeRev: String!
+            ) {
+                discussionThreads(
+                    first: 10000
+                    targetRepositoryGitCloneURL: $targetRepositoryGitCloneURL
+                    targetRepositoryPath: $targetRepositoryPath
+                ) {
+                    totalCount
+                    pageInfo {
+                        hasNextPage
+                    }
+                    nodes {
+                        ...DiscussionThreadFields
+                    }
+                }
+            }
+            ${discussionThreadFieldsFragment}
+        `,
+        {
+            targetRepositoryGitCloneURL: remoteUrl,
+            targetRepositoryPath: path,
+            relativeRev: branch,
+        }
+    )
+
+    if (!data.discussionThreads || !data.discussionThreads.nodes) {
+        throw new Error(`Invalid GraphQL response for DiscussionThreads`)
+    }
+
+    const threads: vscode.CommentThread[] = []
+    for (const thread of data.discussionThreads.nodes) {
+        // TODO: this assumes there is no diff between the document state and the revision
+        const sel = thread.target.relativeSelection
+        if (sel) {
+            const range = new vscode.Range(sel.startLine, sel.startCharacter, sel.endLine, sel.endCharacter)
+            threads.push(discussionToCommentThread(document, range, thread))
+        }
+    }
+
+    return threads
+}
+
+async function createNewCommentThread(
+    document: vscode.TextDocument,
+    range: vscode.Range,
+    text: string
+): Promise<vscode.CommentThread> {
+    log.appendLine(`createNewCommentThread ${document.uri} ${range}`)
+    const [remoteUrl, branch, path] = await repoInfo(document.fileName)
+    if (remoteUrl === '') {
+        throw new Error('Git repository has no remote url configured')
+    }
+
+    const selection = getSelection(document, range)
+
+    const targetRepo: SourcegraphGQL.IDiscussionThreadTargetRepoInput = {
+        repositoryGitCloneURL: remoteUrl,
+        path,
+        branch,
+        selection,
+    }
+    const input: SourcegraphGQL.IDiscussionThreadCreateInput = {
+        title: text,
+        contents: text,
+        targetRepo,
+    }
+
+    const data = await mutateGraphQL(
+        gql`
+            mutation CreateThread($input: DiscussionThreadCreateInput!, $relativeRev: String!) {
+                discussions {
+                    createThread(input: $input) {
+                        ...DiscussionThreadFields
+                    }
+                }
+            }
+            ${discussionThreadFieldsFragment}
+        `,
+        { input, relativeRev: branch }
```

Comment:
`relativeRev` here should be the exact Git revision the user is viewing, not the branch name. The branch name would return the selection relative to the server's current HEAD revision of the named branch which is not want you want and will make the discussion threads randomly placed in the wrong locations in files if your local Git is behind.